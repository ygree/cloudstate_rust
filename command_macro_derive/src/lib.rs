
use proc_macro::TokenStream;
use quote::quote;
use syn;

#[proc_macro_derive(CommandDecoder)]
pub fn command_macro_derive(input: TokenStream) -> TokenStream {
    // Construct a representation of Rust code as a syntax tree
    // that we can manipulate
    let ast = syn::parse(input).unwrap();

    // Build the trait implementation
    impl_command_macro(&ast)
}

fn impl_command_macro(ast: &syn::DeriveInput) -> TokenStream {

    let type_name = &ast.ident;

    let variants = match &ast.data {
        syn::Data::Enum(data_enum) => {
            //TODO check that each variant contains only one unnamed field with a type that implements ::prost::Message trait
            data_enum.variants.iter().map(|v| &v.ident).collect()
        },
        _ => vec![], //TODO return an error that only enums are supported
    };

    let unknown_command = quote! {
        unknown_command_type => {
            eprintln!("Unknown command type: {}", unknown_command_type);
            None
        },
    };

    let items: Vec<_> = variants.iter().map(|v| {
        //TODO Add type package. Need to pass the package name somehow, maybe as an enum attribute?
        //TODO Use internal struct name as a type name for matching instead of the enum variant name!
        let variant_name = v.to_string();
        quote!(
            #variant_name => {
                //TODO use variant param type instead of v here
                match <#v as Message>::decode(bytes) {
                    Ok(cmd) => {
                        println!("Received {:?}", cmd);
                        Some(#type_name::#v(cmd))
                    },
                    Err(err) => {
                        eprintln!("Error decoding {} command: {}", #variant_name, err);
                        None
                    },
                }
            },
        )
    }).collect();

    let gen = quote! {
        impl CommandDecoder for #type_name {
            fn decode(type_url: String, bytes: Bytes) -> Option<Self> {
                match type_url.as_ref() {
                    #(#items)*
                    #unknown_command
                }
            }
        }
    };

    gen.into()
}
