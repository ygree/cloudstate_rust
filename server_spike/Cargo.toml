[package]
name = "server-spike"
version = "0.1.0"
authors = ["Yury Gribkov <yury.gribkov@gmail.com>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
protocols = { path = "../protocols" }
command_macro_derive = { path = "../command_macro_derive" }
tonic = "0.2"
prost = "0.6"
prost-types = "0.6"
tokio = { version = "0.2", features = ["rt-threaded", "time", "stream", "fs", "macros", "uds"] }
futures = { version = "0.3", default-features = false, features = ["alloc"] }
async-stream = "0.2"
tower = "0.3"
serde = { version = "1.0", features = ["derive"] }
serde_json = "1.0"
rand = "0.7"
bytes = "0.5.4"

[[bin]]
name = "spike-server"
path = "src/server.rs"

[[bin]]
name = "spike-client"
path = "src/client.rs"