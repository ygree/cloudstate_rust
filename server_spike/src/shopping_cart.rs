
use ::command_macro_derive::CommandDecoder;
use protocols::example::shoppingcart::{AddLineItem, RemoveLineItem, GetShoppingCart, persistence};
use protocols::example::shoppingcart::persistence::{Cart, ItemAdded, ItemRemoved};
use super::{EventSourcedEntity, CommandDecoder};
use ::prost::Message;
use bytes::Bytes;
use crate::HandleCommandContext;

#[derive(Default)]
pub struct ShoppingCartEntity(Cart);
//TODO use more convenient type for internal state, e.g. HashMap

//TODO consider using Attribute-like macros, e.g. #[commands(AddLineItem, RemoveLineItem, GetShoppingCart)]
// Combine command into one type.
#[derive(CommandDecoder)]
pub enum ShoppingCartCommand {
    AddLineItem(AddLineItem),
    RemoveLineItem(RemoveLineItem),
    GetShoppingCart(GetShoppingCart),
}

pub enum ShoppingCartEvent {
    ItemAdded(ItemAdded),
    ItemRemoved(ItemRemoved),
}

// impl ShoppingCartEvent {
//     fn from_message(msg: impl ::prost::Message) -> Option<ShoppingCartEvent> {
//         match msg {
//             cmd @ ItemAdded {..} => Some(cmd),
//             cmd @ ItemRemoved {..} => Some(cmd),
//             _ => None,
//         }
//     }
// }

impl EventSourcedEntity for ShoppingCartEntity {

    type Snapshot = Cart;
    type Command = ShoppingCartCommand;
    type Event = ShoppingCartEvent;

    fn snapshot_loaded(&mut self, snapshot: Self::Snapshot) {
        self.0 = snapshot;
        println!("Snapshot Loaded: {:?}", self.0);
    }

    fn handle_command(&self, command: Self::Command, context: &mut impl HandleCommandContext<Event=Self::Event>) {
        match command {
            ShoppingCartCommand::AddLineItem(add_line_item) => {
                println!("Handle command: {:?}", add_line_item);
                context.emit_event(
                    //TODO looks like too much boilerplate
                    ShoppingCartEvent::ItemAdded(
                        ItemAdded {
                            item: Some(
                                persistence::LineItem {
                                    product_id: add_line_item.product_id,
                                    name: add_line_item.name,
                                    quantity: add_line_item.quantity,
                                }
                            )
                        }
                    )
                );
            },
            ShoppingCartCommand::RemoveLineItem(remove_line_item) => {
                println!("Handle command: {:?}", remove_line_item);
            },
            ShoppingCartCommand::GetShoppingCart(get_shopping_cart) => {
                println!("Handle command: {:?}", get_shopping_cart);
            },
        }
    }

    fn handle_event(&mut self, event: Self::Event) {
        match event {
            ShoppingCartEvent::ItemAdded(item_added) => {
                println!("Handle event: {:?}", item_added);
                if let Some(item) = item_added.item {
                    self.0.items.push(item);
                }
            },
            ShoppingCartEvent::ItemRemoved(item_removed) => {
                println!("Handle event: {:?}", item_removed);
                //TODO remove item
            },
        }
    }

}

