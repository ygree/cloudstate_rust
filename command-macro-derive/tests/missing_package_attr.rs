use ::prost::Message;
use bytes::Bytes;
use server_spike::CommandDecoder; //TODO should probably move this to a separate trait out of server_spike
use command_macro_derive::CommandDecoder;

mod shopping_cart;
use shopping_cart::*;

// #[package = "com.example.shoppingcart"]
#[derive(CommandDecoder)]
pub enum ShoppingCartCommand {
    AddLine(AddLineItem),
    RemoveLine(RemoveLineItem),
    GetCart(GetShoppingCart),
}

fn main() {}