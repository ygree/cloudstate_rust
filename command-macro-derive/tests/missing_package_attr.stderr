warning: associated type `handleStream` should have an upper camel case name
   --> $DIR/cloudstate.eventsourced.rs:181:14
    |
181 |         type handleStream: Stream<Item = Result<super::EventSourcedStreamOut, tonic::Status>>
    |              ^^^^^^^^^^^^ help: convert the identifier to upper camel case: `HandleStream`
    |
    = note: `#[warn(non_camel_case_types)]` on by default

error: proc-macro derive panicked
  --> $DIR/missing_package_attr.rs:10:10
   |
10 | #[derive(CommandDecoder)]
   |          ^^^^^^^^^^^^^^
   |
   = help: message: Not found package attribute!
